FreeRTOS-Plus/Source/Utilities/mbedtls_freertos/mbedtls_freertos_port.c

```diff
@@ -121,13 +152,37 @@ int mbedtls_platform_recv( void * ctx,
                            size_t len )
 {
     Socket_t socket;
+    BaseType_t recvStatus;
+    int returnStatus = -1;
 
     configASSERT( ctx != NULL );
     configASSERT( buf != NULL );
 
     socket = ( Socket_t ) ctx;
+    recvStatus = FreeRTOS_recv( socket, buf, len, 0 );
+
+    switch( recvStatus )
+    {
+        /* No data could be sent because the socket was or just got closed. */
+        case -pdFREERTOS_ERRNO_ENOTCONN:
+        /* No data could be sent because there was insufficient memory. */
+        case -pdFREERTOS_ERRNO_ENOMEM:
+        /* No data could be sent because xSocket was not a valid TCP socket. */
+        case -pdFREERTOS_ERRNO_EINVAL:
+            returnStatus = MBEDTLS_ERR_SSL_INTERNAL_ERROR;
+            break;
+
+        /* A timeout occurred before any data could be sent. */
+        case -pdFREERTOS_ERRNO_ENOSPC:
+            returnStatus = MBEDTLS_ERR_SSL_TIMEOUT;
+            break;
+
+        default:
+            returnStatus = ( int ) recvStatus;
```

Comment:
What about the case where recvStatus is 0, which I think mbedTLS will take as an error, but might just mean you need to call FreeRTOS_recv() again when data is available?  I'm not sure what will happen here and the consequence might be different depending on whether the read is speculative to find the start of a new frame (in which case if there is no frame we don't want to block and just move on), compared to when a frame is known to be being received (in which case we want to read again).

Reply:
Can we just return `MBEDTLS_ERR_SSL_TIMEOUT` when recvStatus == 0? mbedTLS treats 0 as a special case where the connection is closed. However, if we return `MBEDTLS_ERR_SSL_TIMEOUT` instead, then mbedTLS will treat that as a retriable error.

Reply:
I think if we call `FreeRTOS_recv` again until there is data to receive, the timeout initially passed to `TLS_FreeRTOS_Connect` function no longer applies. After diving into mbedTLS's usage of the callback, it does not seem like it will attempt any speculative reads. It always tries to receive as many bytes as it can from the stream and into the message buffer.

Another advantage with returning `MBEDTLS_ERR_SSL_TIMEOUT` is that it will attempt renegotiation with the server if it is pending.